src/dyn_subgrid/dyncropFileMod.F90

```diff
@@ -165,8 +166,12 @@ subroutine dyncrop_interp(bounds,crop_inst)
     allocate(fertcft_cur(bounds%begg:bounds%endg, cft_lb:cft_ub))
     call fertcft%get_current_data(fertcft_cur)
 
-    if (use_crop) then
-       call collapse_crop_types(wtcft_cur, fertcft_cur, bounds%begg, bounds%endg, verbose = .false.)
+    ! Call collapse_crop_types when we read 78-pft datasets (ie cft_size > 2),
+    ! while running with use_crop = .false.
+    ! This call also requires create_crop_landunit = .true.
+    ! Same if-statement and call appear in surfrdMod.F90.
+    if (.not. use_crop .and. create_crop_landunit .and. cft_size > 2) then
+       call collapse_crop_types(wtcft_cur, fertcft_cur, cft_size, bounds%begg, bounds%endg, verbose = .false.)
```

Comment:
This doesn't look like the right logic to me. This might be complex enough that it should have a unit test to make sure it's right. I think you want this to be done when crop is on (so that you collapse some of the crop types onto each other), but also for the case when crop is off and cft_size > 2.

Reply:
You're probably right, and I think this means that I had it right the first time when I just checked for cft_size > 2...

I will look into adding a unit test and will ask for help if I get lost.

Reply:
I can't envision what @ekluzek is thinking here in terms of a unit test. I think this is going to be very hard to unit test as is (i.e., with the logic embedded in this routine as it is now), and I probably wouldn't go there - at least, not as the first unit tests you write for ctsm. @ekluzek were you thinking that the conditional could be extracted to a little standalone logical function like 'cft_needs_collapse'? I was tempted to suggest something like that, but that mainly seems to make sense if the logic is shared between surfrdMod and this file, which based on your comments, sounds like it may not be.

So if the logic ends up differing between the two, I'll just reiterate my original comment that I'd like to see some comments explaining why the logic is what it is (and why it differs from the logic with a similar purpose in surfrdMod, if relevant).

Reply:
The test suite flags the last change as a problem, producing FAILures such as:
ERP_Ly3_P72x2.f10_f10_musgs.IHistClm50BgcCrop.cheyenne_intel.clm-cropMonthOutput
ERP_D_Ld5.f10_f10_musgs.IHistClm50BgcCrop.cheyenne_intel.clm-allActive

I will return to just checking for cft_size > 2 and will modify the corresponding comment.

Knowing that the test-suite FAILs, is it still necessary to also make a unit test? If so, what do you have in mind?

Reply:
I guess what I was envisioning is making it a function call that you then test. And my main point is that complex logic always tends to be something that's easily incorrect. So doing testing on complex logic is a good thing. But, really what you want to make sure of is that it's being called when it should. One check for that, for example, is we want to make sure that crop simulations don't have a change in answers. So the system tests can verify that. Another approach to this is to call it more often than you should as long as it doesn't do the wrong thing. I think it's possible that collapse_crop_types call would in some cases do something important, but in others essentially amount to doing nothing, so it's OK to call it too much -- just not too little.

Reply:
@ekluzek I was starting to think along similar lines as you (I think) – specifically, thinking that we should just call collapse_crop_types all the time, and just let it be a no-op if you're already using a dataset CFTs that don't need collapsing (e.g., a 2-cft dataset). However, if I'm reading `collapse_crop_types` correctly, it looks hard-coded to assume that the inputs have CFTs that go up to npcropmax (hard-coded at 78). My intuitive feeling is that what should really be done here is to generalize `collapse_crop_types` so we can just call it always and have it Just Work. This routine is a perfect candidate for unit testing (maybe that's already done), so we can throw a variety of scenarios at it to make sure they're handled properly. But until that's done, I think we need a conditional around the call to collapse_crop_types.